ta/crypto_perf/ta_crypto_perf.c

```diff
@@ -465,8 +467,476 @@ TEE_Result cmd_hash_prepare_op(uint32_t param_types, TEE_Param params[4])
+						   TEE_PARAM_TYPE_VALUE_INPUT,
+						   TEE_PARAM_TYPE_MEMREF_INPUT,
+						   TEE_PARAM_TYPE_MEMREF_OUTPUT);
+	TEE_Attribute *attrs = NULL;
+	uint32_t attr_count = 0;
+	TEE_Result (*do_asym)(TEE_OperationHandle, const TEE_Attribute *,
+			      uint32_t, const void *, uint32_t, void *,
+			      uint32_t *) = NULL;
+
+	if (param_types != exp_param_types)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	res = unpack_attrs(params[0].memref.buffer, params[0].memref.size,
+			   &attrs, &attr_count);
+	if (res != TEE_SUCCESS)
+		return res;
+
+	n = params[1].value.a;
+	mode = params[1].value.b;
+
+	if (mode == MODE_ENCRYPT)
+		do_asym = TEE_AsymmetricEncrypt;
+	else if (mode == MODE_DECRYPT)
+		do_asym = TEE_AsymmetricDecrypt;
+	else if (mode == MODE_SIGN)
+		do_asym = TEE_AsymmetricSignDigest;
+	else if (mode == MODE_VERIFY)
+		do_asym = TEE_AsymmetricVerifyDigest;
+
+	while (n--) {
+		res = do_asym(crypto_op, attrs, attr_count,
+			      params[2].memref.buffer, params[2].memref.size,
+			      params[3].memref.buffer, &params[3].memref.size);
+
+		CHECK(res, "TEE processing failed", goto out;);
+	}
+
+out:
+	TEE_Free(attrs);
+	return res;
+}
+
+TEE_Result cmd_asym_prepare_obj(uint32_t param_types,
+				TEE_Param params[TEE_NUM_PARAMS])
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+	uint32_t tee_type = TEE_TYPE_ILLEGAL_VALUE;
+	uint32_t exp_param_types = TEE_PARAM_TYPES(TEE_PARAM_TYPE_VALUE_INPUT,
+						   TEE_PARAM_TYPE_NONE,
+						   TEE_PARAM_TYPE_NONE,
+						   TEE_PARAM_TYPE_NONE);
+
+	if (param_types != exp_param_types)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	tee_type = get_keypair_type(params[0].value.a);
+	if (tee_type == TEE_TYPE_ILLEGAL_VALUE)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	cmd_clean_obj();
+	res = TEE_AllocateTransientObject(tee_type, params[0].value.b,
+					  &crypto_obj);
+	CHECK(res, "TEE_AllocateTransientObject", return res;);
+
+	return TEE_SUCCESS;
+}
+
+TEE_Result cmd_asym_prepare_keypair(uint32_t param_types,
+				    TEE_Param params[TEE_NUM_PARAMS])
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+	TEE_Attribute *attrs = NULL;
+	uint32_t attr_count = 0;
+	uint32_t width_bits = 0;
+	uint32_t algo_type = 0;
+	uint32_t tee_type = 0;
+	uint32_t mode = 0;
+	uint32_t exp_param_types = TEE_PARAM_TYPES(TEE_PARAM_TYPE_VALUE_INPUT,
+						   TEE_PARAM_TYPE_VALUE_INPUT,
+						   TEE_PARAM_TYPE_MEMREF_INPUT,
+						   TEE_PARAM_TYPE_NONE);
+
+	if (param_types != exp_param_types)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	mode = params[0].value.b;
+	width_bits = params[0].value.a;
+	tee_type = params[1].value.a;
+	algo_type = params[1].value.b;
+
+	if (get_algo(tee_type, mode, width_bits, algo_type))
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	res = unpack_attrs(params[2].memref.buffer, params[2].memref.size,
+			   &attrs, &attr_count);
+	if (res != TEE_SUCCESS)
+		return res;
+
+	res = TEE_GenerateKey(crypto_obj, width_bits, attrs, attr_count);
+	CHECK(res, "TEE_GenerateKey", goto out;);
```

Comment:
s/TEE_GenerateKey/TEE_GenerateKey()/ to highlight it's a functin label.

Ditto at lines 827, 830, 835, 838, 842, 845 as well as at line 874, 880 and 786, 708.